File: penguin/cli/command_registry.py
"""
Command Registry System for Penguin TUI.

Loads and manages commands from commands.yml configuration.
"""

import yaml
import os
from pathlib import Path
from typing import Dict, List, Optional, Any, Callable, Tuple
from dataclasses import dataclass, field
import logging

logger = logging.getLogger(__name__)


@dataclass
class CommandParameter:
    """Parameter definition for a command."""
    name: str
    type: str
    required: bool = True
    description: str = ""
    default: Any = None


@dataclass 
class Command:
    """Represents a TUI command."""
    name: str
    category: str
    description: str
    handler: str
    aliases: List[str] = field(default_factory=list)
    parameters: List[CommandParameter] = field(default_factory=list)
    enabled: bool = True
    
    def matches(self, input_str: str) -> bool:
        """Check if input matches this command or its aliases."""
        if input_str == self.name:
            return True
        return input_str in self.aliases
    
    def parse_args(self, args_str: str) -> Dict[str, Any]:
        """Parse arguments string into parameter dict."""
        import shlex
        args: Dict[str, Any] = {}

        # No arguments provided
        if not args_str or not args_str.strip():
            for param in self.parameters:
                if not param.required:
                    args[param.name] = param.default
            return args

        # Use shlex to honor quotes
        try:
            parts = shlex.split(args_str)
        except ValueError:
            parts = args_str.split()

        for i, param in enumerate(self.parameters):
            if i < len(parts):
                raw = parts[i]
                if param.type == "int":
                    try:
                        args[param.name] = int(raw)
                    except ValueError:
                        args[param.name] = param.default
                elif param.type == "bool":
                    args[param.name] = str(raw).lower() in ("true", "yes", "1", "on")
                else:
                    args[param.name] = raw
            elif not param.required:
                args[param.name] = param.default
            # else: leave missing required param unfilled

        return args


class CommandRegistry:
    """
    Registry for TUI commands loaded from YAML configuration.
    
    Features:
    - Load commands from commands.yml
    - Support for aliases
    - Command categorization
    - Plugin extension support (future)
    - MCP integration (future)
    """
    
    def __init__(self, config_path: Optional[str] = None):
        self.commands: Dict[str, Command] = {}
        self.aliases: Dict[str, str] = {}
        self.categories: Dict[str, List[Command]] = {}
        self.handlers: Dict[str, Callable] = {}
        
        # Load configuration
        if config_path:
            self.config_path = Path(config_path)
        else:
            # Default to commands.yml in the same directory as this file
            self.config_path = Path(__file__).parent / "commands.yml"
        
        self.load_commands()
    
    def load_commands(self) -> None:
        """Load commands from YAML configuration."""
        if not self.config_path.exists():
            logger.warning(f"Commands config not found: {self.config_path}")
            self._register_builtin_commands()
            return
        
        try:
            with open(self.config_path, 'r') as f:
                config = yaml.safe_load(f)
            
            # Process categories
            for cat_def in config.get('categories', []):
                self.categories[cat_def['name']] = []
            
            # Process commands
            for cmd_def in config.get('commands', []):
                command = self._create_command(cmd_def)
                if command.enabled:
                    self.register(command)
            
            logger.info(f"Loaded {len(self.commands)} commands from {self.config_path}")
            
        except Exception as e:
            logger.error(f"Error loading commands config: {e}")
            self._register_builtin_commands()
    
    def _create_command(self, cmd_def: Dict[str, Any]) -> Command:
        """Create a Command from YAML definition."""
        # Parse parameters
        params = []
        for param_def in cmd_def.get('parameters', []):
            param = CommandParameter(
                name=param_def['name'],
                type=param_def.get('type', 'string'),
                required=param_def.get('required', True),
                description=param_def.get('description', ''),
                default=param_def.get('default')
            )
            params.append(param)
        
        return Command(
            name=cmd_def['name'],
            category=cmd_def.get('category', 'general'),
            description=cmd_def.get('description', ''),
            handler=cmd_def.get('handler', ''),
            aliases=cmd_def.get('aliases', []),
            parameters=params,
            enabled=cmd_def.get('enabled', True)
        )
    
    def register(self, command: Command) -> None:
        """Register a command."""
        self.commands[command.name] = command
        
        # Register aliases
        for alias in command.aliases:
            self.aliases[alias] = command.name
        
        # Add to category
        if command.category not in self.categories:
            self.categories[command.category] = []
        self.categories[command.category].append(command)
    
    def register_handler(self, handler_name: str, handler: Callable) -> None:
        """Register a command handler function."""
        self.handlers[handler_name] = handler
    
    def find_command(self, input_str: str) -> Optional[Command]:
        """Find command by exact name or alias (no partial matching)."""
        if input_str in self.commands:
            return self.commands[input_str]
        if input_str in self.aliases:
            return self.commands[self.aliases[input_str]]
        return None
    
    def parse_input(self, input_str: str) -> Tuple[Optional[Command], Dict[str, Any]]:
        """
        Parse user input into command and arguments.
        
        Returns:
            Tuple of (Command, arguments dict) or (None, {}) if not found
        """
        input_str = input_str.strip()
        if not input_str:
            return None, {}
        
        # Remove leading slash if present
        if input_str.startswith('/'):
            input_str = input_str[1:]
        
        # Tokenize input
        parts = input_str.split()
        command: Optional[Command] = None
        remaining_args = ""

        # 1) Exact match on full string first
        cmd_candidate = input_str
        if cmd_candidate in self.commands or cmd_candidate in self.aliases:
            command = self.find_command(cmd_candidate)
            if command:
                used_tokens = len(command.name.split())
                remaining_args = " ".join(parts[used_tokens:])

        # 2) Longest prefix match by command tokens
        if not command:
            best_name = None
            best_len = -1
            for cmd_name in self.commands.keys():
                cmd_tokens = cmd_name.split()
                if len(parts) >= len(cmd_tokens) and parts[:len(cmd_tokens)] == cmd_tokens:
                    if len(cmd_tokens) > best_len:
                        best_name = cmd_name
                        best_len = len(cmd_tokens)
            if best_name:
                command = self.commands[best_name]
                remaining_args = " ".join(parts[best_len:])

        if not command:
            return None, {}

        # Parse arguments from the remaining string
        try:
            args = command.parse_args(remaining_args) if remaining_args else {}
        except ValueError as e:
            logger.error(f"Error parsing command arguments: {e}")
            args = {}
        
        return command, args
    
    def get_suggestions(self, partial: str) -> List[str]:
        """Get command suggestions for autocomplete."""
        if not partial:
            return []
        
        # Remove leading slash
        if partial.startswith('/'):
            partial = partial[1:]
        
        suggestions: List[str] = []
        
        # Check direct commands
        for cmd_name in self.commands:
            if cmd_name.startswith(partial):
                suggestions.append(f"/{cmd_name}")
        
        # Check aliases
        for alias in self.aliases:
            if alias.startswith(partial):
                suggestions.append(f"/{alias}")
        
        return sorted(suggestions)[:10]  # Limit to 10 suggestions
    
    def get_help_text(self) -> str:
        """Generate help text for all commands."""
        lines = ["**Available Commands:**\n"]
        
        for category, commands in sorted(self.categories.items()):
            if not commands:
                continue
            
            # Category header
            cat_title = category.replace('_', ' ').title()
            lines.append(f"\n**{cat_title}:**")
            
            # Commands in category
            for cmd in sorted(commands, key=lambda c: c.name):
                # Command with aliases
                cmd_str = f"/{cmd.name}"
                if cmd.aliases:
                    alias_str = ", ".join(f"/{a}" for a in cmd.aliases)
                    cmd_str += f" ({alias_str})"
                
                lines.append(f"- `{cmd_str}` - {cmd.description}")
                
                # Parameters
                if cmd.parameters:
                    for param in cmd.parameters:
                        req = "required" if param.required else "optional"
                        lines.append(f"    • {param.name} ({param.type}, {req}): {param.description}")
        
        return "\n".join(lines)
    
    def _register_builtin_commands(self) -> None:
        """Register minimal builtin commands as fallback."""
        builtins = [
            Command("help", "system", "Show help", "_show_help", ["h", "?"]),
            Command("clear", "chat", "Clear chat", "action_clear_log", ["cls"]),
            Command("quit", "system", "Exit", "action_quit", ["exit", "q"]),
            Command("debug", "debug", "Show debug info", "action_show_debug"),
        ]
        
        for cmd in builtins:
            self.register(cmd)

File: misc/penguin-cli/src/core/commands/CommandRegistry.ts
/**
 * Command Registry System for Penguin Ink CLI
 *
 * Loads and manages commands from commands.yml configuration.
 * Port of command_registry.py to TypeScript.
 */

import yaml from 'js-yaml';
import fs from 'fs';
import path from 'path';
import { logger } from '../../utils/logger';
import type {
  Command,
  CommandCategory,
  CommandConfig,
  CommandParameter,
  ParsedCommand,
  CommandHandler,
  CommandHandlers,
} from './types.js';

export class CommandRegistry {
  private commands: Map<string, Command> = new Map();
  private aliases: Map<string, string> = new Map();
  private categories: Map<string, CommandCategory> = new Map();
  private handlers: CommandHandlers = {};

  constructor(private configPath?: string) {
    if (!configPath) {
      // Default to commands.yml in the penguin/cli directory
      this.configPath = path.join(process.cwd(), '..', 'penguin', 'cli', 'commands.yml');
    }
    this.loadCommands();
  }

  /**
   * Load commands from YAML configuration file.
   */
  private loadCommands(): void {
    // Always register builtin commands first
    this.registerBuiltinCommands();

    if (!this.configPath || !fs.existsSync(this.configPath)) {
      logger.debug(`[CommandRegistry] Commands config not found: ${this.configPath}`);
      return;
    }

    try {
      const fileContent = fs.readFileSync(this.configPath, 'utf8');
      const config = yaml.load(fileContent) as CommandConfig;

      // Process categories
      if (config.categories) {
        for (const category of config.categories) {
          this.categories.set(category.name, category);
        }
      }

      // Process commands (these can override builtins)
      if (config.commands) {
        for (const cmdDef of config.commands) {
          const command = this.createCommand(cmdDef);
          if (command.enabled !== false) {  // Changed from command.enabled to handle undefined
            this.register(command);
          }
        }
      }

      logger.debug(`[CommandRegistry] Loaded ${this.commands.size} commands (including builtins)`);
      // Debug: Show first few commands
      const commandNames = Array.from(this.commands.keys()).slice(0, 15);
      logger.debug(`[CommandRegistry] Sample commands: ${commandNames.join(', ')}`);
    } catch (error) {
      logger.warn(`[CommandRegistry] Error loading commands config:`, error as Error);
    }
  }

  /**
   * Create a Command object from YAML definition.
   */
  private createCommand(cmdDef: any): Command {
    const parameters: CommandParameter[] = [];

    if (cmdDef.parameters) {
      for (const paramDef of cmdDef.parameters) {
        parameters.push({
          name: paramDef.name,
          type: paramDef.type || 'string',
          required: paramDef.required ?? true,
          description: paramDef.description || '',
          default: paramDef.default,
        });
      }
    }

    return {
      name: cmdDef.name,
      category: cmdDef.category || 'general',
      description: cmdDef.description || '',
      handler: cmdDef.handler || '',
      aliases: cmdDef.aliases || [],
      parameters,
      enabled: cmdDef.enabled ?? true,
    };
  }

  /**
   * Register a command in the registry.
   */
  register(command: Command): void {
    this.commands.set(command.name, command);

    // Register aliases
    for (const alias of command.aliases) {
      this.aliases.set(alias, command.name);
    }
  }

  /**
   * Register a command handler function.
   */
  registerHandler(handlerName: string, handler: CommandHandler): void {
    this.handlers[handlerName] = handler;
  }

  /**
   * Find command by exact name or alias.
   */
  findCommand(input: string): Command | null {
    if (this.commands.has(input)) {
      return this.commands.get(input)!;
    }
    if (this.aliases.has(input)) {
      const commandName = this.aliases.get(input)!;
      return this.commands.get(commandName)!;
    }
    return null;
  }

  /**
   * Parse user input into command and arguments.
   *
   * Returns ParsedCommand or null if not a valid command.
   */
  parseInput(input: string): ParsedCommand | null {
    let trimmed = input.trim();
    if (!trimmed) return null;

    // Remove leading slash if present
    if (trimmed.startsWith('/')) {
      trimmed = trimmed.slice(1);
    }

    const parts = trimmed.split(/\s+/);
    let command: Command | null = null;
    let remainingArgs = '';

    // 1) Try exact match on full string
    if (this.commands.has(trimmed) || this.aliases.has(trimmed)) {
      command = this.findCommand(trimmed);
      if (command) {
        const usedTokens = command.name.split(/\s+/).length;
        remainingArgs = parts.slice(usedTokens).join(' ');
      }
    }

    // 2) Try longest prefix match (for multi-word commands like "chat list")
    if (!command) {
      let bestName: string | null = null;
      let bestLen = -1;

      for (const [cmdName] of this.commands) {
        const cmdTokens = cmdName.split(/\s+/);
        if (parts.length >= cmdTokens.length) {
          const match = cmdTokens.every((token, i) => parts[i] === token);
          if (match && cmdTokens.length > bestLen) {
            bestName = cmdName;
            bestLen = cmdTokens.length;
          }
        }
      }

      if (bestName) {
        command = this.commands.get(bestName)!;
        remainingArgs = parts.slice(bestLen).join(' ');
      }
    }

    if (!command) return null;

    // Parse arguments
    const args = this.parseArgs(command, remainingArgs);
    return { command, args };
  }

  /**
   * Parse arguments string into parameter dict.
   */
  private parseArgs(command: Command, argsStr: string): Record<string, any> {
    const args: Record<string, any> = {};

    if (!argsStr || !argsStr.trim()) {
      // No arguments provided - use defaults
      for (const param of command.parameters) {
        if (!param.required) {
          args[param.name] = param.default;
        }
      }
      return args;
    }

    // Simple split by whitespace (could use shlex-like parsing later)
    const parts = argsStr.trim().split(/\s+/);

    for (let i = 0; i < command.parameters.length; i++) {
      const param = command.parameters[i];
      const isLastParam = i === command.parameters.length - 1;

      if (i < parts.length) {
        // For the last parameter, capture all remaining text
        const raw = isLastParam ? parts.slice(i).join(' ') : parts[i];

        // Type conversion
        switch (param.type) {
          case 'int':
            const parsed = parseInt(raw, 10);
            args[param.name] = isNaN(parsed) ? param.default : parsed;
            break;
          case 'bool':
            args[param.name] = ['true', 'yes', '1', 'on'].includes(raw.toLowerCase());
            break;
          default:
            args[param.name] = raw;
        }
      } else if (!param.required) {
        args[param.name] = param.default;
      }
    }

    return args;
  }

  /**
   * Get command suggestions for autocomplete.
   */
  getSuggestions(partial: string): string[] {
    if (!partial) return [];

    // Remove leading slash
    const search = partial.startsWith('/') ? partial.slice(1) : partial;
    const suggestions: string[] = [];

    // Check commands
    for (const [cmdName] of this.commands) {
      if (cmdName.startsWith(search)) {
        suggestions.push(`/${cmdName}`);
      }
    }

    // Check aliases
    for (const [alias] of this.aliases) {
      if (alias.startsWith(search)) {
        suggestions.push(`/${alias}`);
      }
    }

    return suggestions.sort().slice(0, 10); // Limit to 10
  }

  /**
   * Get help text for all commands.
   */
  getHelpText(): string {
    const lines: string[] = ['**Available Commands:**\n'];

    // Group commands by category
    const byCategory = new Map<string, Command[]>();
    for (const [, command] of this.commands) {
      if (!byCategory.has(command.category)) {
        byCategory.set(command.category, []);
      }
      byCategory.get(command.category)!.push(command);
    }

    // Sort categories and commands
    const sortedCategories = Array.from(byCategory.keys()).sort();

    for (const categoryName of sortedCategories) {
      const commands = byCategory.get(categoryName)!;
      if (commands.length === 0) continue;

      // Category header
      const category = this.categories.get(categoryName);
      const icon = category?.icon || '';
      const title = categoryName.replace(/_/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase());
      lines.push(`\n**${icon} ${title}:**`);

      // Commands in category
      for (const cmd of commands.sort((a, b) => a.name.localeCompare(b.name))) {
        let cmdStr = `/${cmd.name}`;
        if (cmd.aliases.length > 0) {
          const aliasStr = cmd.aliases.map((a) => `/${a}`).join(', ');
          cmdStr += ` (${aliasStr})`;
        }

        lines.push(`- \`${cmdStr}\` - ${cmd.description}`);

        // Parameters
        if (cmd.parameters.length > 0) {
          for (const param of cmd.parameters) {
            const req = param.required ? 'required' : 'optional';
            lines.push(`    • ${param.name} (${param.type}, ${req}): ${param.description}`);
          }
        }
      }
    }

    return lines.join('\n');
  }

  /**
   * Execute a command by name with arguments.
   */
  async execute(commandName: string, args: Record<string, any>, context?: any): Promise<void> {
    const command = this.findCommand(commandName);
    if (!command) {
      throw new Error(`Command not found: ${commandName}`);
    }

    const handler = this.handlers[command.handler];
    if (!handler) {
      throw new Error(`Handler not found: ${command.handler} for command ${commandName}`);
    }

    await handler(args, context);
  }

  /**
   * Register minimal builtin commands as fallback.
   */
  private registerBuiltinCommands(): void {
    const builtins: Command[] = [
      {
        name: 'help',
        category: 'system',
        description: 'Show help',
        handler: '_show_help',
        aliases: ['h', '?'],
        parameters: [],
        enabled: true,
      },
      {
        name: 'clear',
        category: 'chat',
        description: 'Clear chat',
        handler: 'action_clear_log',
        aliases: ['cls'],
        parameters: [],
        enabled: true,
      },
      {
        name: 'quit',
        category: 'system',
        description: 'Exit',
        handler: 'action_quit',
        aliases: ['exit', 'q'],
        parameters: [],
        enabled: true,
      },
      {
        name: 'config edit',
        category: 'configuration',
        description: 'Open config file in $EDITOR',
        handler: 'config_edit',
        aliases: [],
        parameters: [],
        enabled: true,
      },
      {
        name: 'config check',
        category: 'configuration',
        description: 'Validate configuration',
        handler: 'config_check',
        aliases: [],
        parameters: [],
        enabled: true,
      },
      {
        name: 'config debug',
        category: 'configuration',
        description: 'Show diagnostic information',
        handler: 'config_debug',
        aliases: [],
        parameters: [],
        enabled: true,
      },
      {
        name: 'image',
        category: 'workflow',
        description: 'Attach an image to your message',
        handler: 'attach_image',
        aliases: ['img'],
        parameters: [
          {
            name: 'path',
            type: 'string',
            required: true,
            description: 'Path to the image file',
          },
          {
            name: 'message',
            type: 'string',
            required: false,
            description: 'Optional message about the image',
          },
        ],
        enabled: true,
      },
    ];

    for (const command of builtins) {
      this.register(command);
    }
  }

  /**
   * Get all registered commands.
   */
  getAllCommands(): Command[] {
    return Array.from(this.commands.values());
  }

  /**
   * Get all categories.
   */
  getAllCategories(): CommandCategory[] {
    return Array.from(this.categories.values());
  }
}
